helper/resource/testing.go

```diff
@@ -959,6 +960,34 @@ func testCheckResourceAttr(is *terraform.InstanceState, name string, key string,
 	return nil
 }
 
+// CheckResourceAttrWithFunc is the callback type used to apply a custom checking logic
+// when using TestCheckResourceAttrWith and a value is found for the given name and key.
+//
+// When this function returns an error, TestCheckResourceAttrWith will fail the check.
+type CheckResourceAttrWithFunc func(value string) error
+
+// TestCheckResourceAttrWith - as per TestCheckResourceAttr but the attribute value
+// checking logic can be customised.
+//
+// If a value is found for the given name and key, it is passed to the CheckResourceAttrWithFunc function.
+// The CheckResourceAttrWithFunc function can then apply any checking logic:
+// should return an error for the check to fail, or `nil` to succeed.
```

Comment:
I like this description here, I'm just wondering if we should also give provider developers all the usage information when they reference it in Go documentation or via the Go language server (e.g. hovering in an editor), similar to the other `TestCheckFunc`

```go
// TestCheckResourceAttrWith ensures a specific value is stored in state for
// the given name and key combination, by checking it against a custom
// validation function. State value checking is only recommended for testing
// Computed attributes and attribute defaults.
//
// For managed resources, the name parameter is combination of the resource
// type, a period (.), and the name label. The name for the below example
// configuration would be "myprovider_thing.example".
//
//     resource "myprovider_thing" "example" { ... }
//
// For data sources, the name parameter is a combination of the keyword "data",
// a period (.), the data source type, a period (.), and the name label. The
// name for the below example configuration would be
// "data.myprovider_thing.example".
//
//     data "myprovider_thing" "example" { ... }
//
// The key parameter is an attribute path in Terraform CLI 0.11 and earlier
// "flatmap" syntax. Keys start with the attribute name of a top-level
// attribute. Use the following special key syntax to inspect list, map, and
// set attributes:
//
//     - .{NUMBER}: List value at index, e.g. .0 to inspect the first element.
//        Use the TestCheckTypeSet* and TestMatchTypeSet* functions instead
//        for sets.
//     - .{KEY}: Map value at key, e.g. .example to inspect the example key
//        value.
//
// The f parameter is a function containing the custom value validation logic.
// Any returned errors from that function will cause the check and test to fail.
```

Reply:
Makes sense. I'll update.

Reply:
I have provided a much more comprehensive godoc for this function.

Let me know your thoughts